### ISSUE TITLE
Issues with global roots management

### ISSUE THREAD

**Issue by kayceesrk** [ocamllabs/ocaml-multicore#424]
There are several issues currently with global roots management. I thought it might be best to post a single issue that gathers everything together and then make separate issues for each of them. 

## Global roots

OCaml allows the creation of global roots, which are considered to be roots for the GC in addition to the current stack, the live registers, and the C local roots. The global roots in Multicore OCaml come in 4 variants. These variants share some of the implementations, but it is good to classify them separately for ease of understanding. 

### 1: Roots used by the compiler itself

`caml_create_root`, `caml_delete_root`, `caml_read_root`, `caml_modify_root`.

These roots are initialised with `caml_create_root` which returns a `caml_root` value. The roots can be read and deleted. The root must be modified with `caml_modify_root`. This is Multicore OCaml only, and is meant to operate nicely with the concurrent GC. 

### 2: Global roots referred to by C global variables

`caml_register_global_root`, `caml_remove_global_root`

This API allows registering C global variables of type `value` to be considered as a root. In particular, the global variable can be modified directly and the GC ensures that the new value remains reachable. 

### 3: Global roots referred to by C global variables but not modified directly 

`caml_register_generational_global_root`, `caml_remove_generational_global_root`, `caml_modify_generational_global_root`

The global variable must not be modified directly but must go through `caml_modify_generational_global_root`. This is done for performance.

### 4: Native dynlinked globals

`caml_register_dyn_global`

### 5: caml_globals created by the native compiler

These globals are created by the `asmcomp` compiler and do not have an API. There is a global table of such globals named `caml_globals` that is accessed by the runtime. 

## Implementation

(1) is implemented by having a linked list `roots_all` that is a valid OCaml object, whose spine is in the major heap, and each value is the registered root either in the major or the minor heap. The clever trick is to that `caml_modify_root` uses write barrier to update the root and hence, the usual inter-generational pointer mechanism kicks in for making sure that the root is kept alive. The linked list of roots is only scanned during the major GC. 

Having `roots_all` appear as a regular OCaml linked list has advantages for the concurrent mark-sweep collection. At the beginning of the major cycle, all of the domains try to mark the head of `roots_all` linked list. In the common case, one of the domains will mark it and push the head into its mark stack. The rest of the domains will see the head of `roots_all` marked and will not push it into their mark stacks. Either way, the roots are safely marked by concurrent domains without much wasted work. 

(2) currently builds on (1). Whenever a root is registered with `caml_register_global_root`, a linked list entry:

```c
/* linked list of registered global roots */                                        
typedef struct capi_global_roots {                                                  
  void *v;                                                                          
  caml_root root;                                                                   
  struct capi_global_roots *next;                                                   
} capi_global_roots; 
```

is created, where `v` holds the address of the global. A new root `root` is created using `caml_create_root` which holds the value of `*v`, the actual root. Since the roots in (2) can be directly updated rather than going through `caml_modify_root`. On modifying the root directly, `*v` will hold the latest value, whereas `root` will continue to point to the old value. **Issue 1**

Worse, this code goes wrong even without modification. Consider that `*v` is initially in the minor heap when the root is registered. On minor GC, that value is promoted to the major heap since it is registered using `caml_create_root`. But `*v` continues to point to the old minor heap location. **Issue 2**

(3) is implemented directly by calling (2) without the optimisations put in by stock OCaml. Since (2) is broken, (3) is broken. **Issue 3**

(4) and (5) have a performance issues. All the domains iterate over all the roots in (4) and (5). It is sufficient if one domain does this work. See `scan_native_globals`. **Issue 4**

When roots are deleted using `caml_delete_root`, the corresponding linked list entry is marked for deletion, but are not removed; the linked list is singly linked and deletion is expensive. There is a function in `globroots.c` named `caml_cleanup_deleted_roots`, which does the cleanup. But this function is never called. Hence, the linked list nodes are never removed. **Issue 5**

## Summary

Now that multicore does not do concurrent minor collection and does stop-the-world parallel collection, it might be useful to revert the changes back to what stock OCaml does, with optimisation for making sure that only the first domain that reaches the scanning functions does the work.

**Comment by sadiqj** [comment]
https://github.com/ocaml-multicore/ocaml-multicore/pull/426 includes some preliminary fixes and also pulls in the gc roots tests from trunk.

The consensus seems to be that we back (1) on to (3) and then remove the `caml_root` API entirely, which would take us to effectively trunk's implementation with some extra locking.

On the performance side, only one domain needs to actually scan the global roots in a minor collection, we can achieve this using the current minor collection count and a try_lock.

For the major collection we need to separate out those roots that are tracked with `caml_register_global_root` (2) and the generational roots (3). We must make sure that all of (2) are scanned by all mutators at the start of a cycle in a stw pause. Those in (3) can be scanned outside of a stop the world pause as they have a deletion barrier.

**Comment by kayceesrk** [comment]
I had a chat with @Sudha247 and she will continue working on the rest of the items on this list.

**Comment by Sudha247** [comment]
This is an update on the current status of global roots management and some possible options to consider for performance improvement. Correctness issues have been resolved by #426.

## Benchmarks

Some micro benchmarks pertaining to global roots have been added to Sandmark in https://github.com/ocaml-bench/sandmark/pull/200. I'm in the process of adding more synthetic benchmarks [here](https://github.com/Sudha247/code-samples/tree/master/multicore-gcroots). These benchmark results together with Sandmark's standard macro benchmarks and their eventlogs have been used to draw some conclusions regarding performance and will be used in future.

## Major Collector

All mutators during a major collection cycle go through all global roots. There is no issue of correctness here, but multiple domains trying to scan the same object is work that can be skipped. Some potential ways to tackle this are listed below.

* **Scanning from one domain:** The global roots can be scanned by exclusively one domain. This needs to be done inside a stop-the-world section, before other domains start marking. This is possibly better than all domains scanning, though it could translate to one domain having more marking work compared with the others.

* **Statically sharing roots to be scanned:** Also inside a stop-the-world pause, the global roots can be statically shared among the available domains. This should be useful if there are a large number of global roots. If the number of roots is small, the overhead of splitting may well be more than the benefits of sharing. If that happens, a fast path of one domain scanning the roots can be done. I think it is worth implementing this option and see the performance results.

* **Incermental marking:** Trunk OCaml marks native globals incrementally, this is also something we can explore. Incremental marking may reduce pausetimes. Some more analysis on benchmark results is required to find out if it will be useful to adapt this to multicore.

## Minor Collector

Earlier, all domains in a stop-the-world pause were going through all young roots. After #463, the first one to reach the scanning part will go through the young roots. This has eliminated other domains going through roots that have already been scanned. The micro-benchmark results have been slightly favourable, while there isn't much changes observed in macro benchmarks' performance. 

We could share this work to other domains, similar to what's mentioned for the major collector, but it is not clear whether this will give any benefits, given that there are going to be overheads added by the sharing process itself. My current understanding is, sharing could be beneficial in cases where there are a large number of young roots, but may not be so for others, which are a majority. As per some general observations from the micro and macro benchmarks, global root scanning does not take up a large part of minor collection, so it might be okay to go with the current implementation for now.

## Removing `caml_root` API

There's an open PR #459 which aims to do this. As mentioned above, the idea is to take our implementation closer to trunk OCaml. The PR has some issues with stack allocated values that needs to be fixed, which I shall be working on, after which it should be ready for review.

---

To summarise, statically sharing global roots scanning in major collection seems worth implementing and minor collection can be unchanged for now. Any ideas and benchmarks related to this are welcome!

**Comment by lpw25** [comment]
> Trunk OCaml marks native globals incrementally, this is also something we can explore. Incremental marking may reduce pausetimes. Some more analysis on benchmark results is required to find out if it will be useful to adapt this to multicore.

You should check with @damiendoligez because I could be mis-remembering this. But I believe that [he implemented incremental scanning of global roots](https://github.com/ocaml/ocaml/pull/297) based on examining the latencies of some of our production systems, and that it was a measurable improvement.

**Comment by damiendoligez** [comment]
Indeed, you remember correctly.

**Comment by Sudha247** [comment]
>based on examining the latencies of some of our production systems, and that it was a measurable improvement.

There is a set of synthetic global roots micro-benchmarks added to sandmark [here](https://github.com/ocaml-bench/sandmark/tree/master/benchmarks/multicore-gcroots). We intend to use those, along with some GC intensive macro benchmarks we have on sandmark (e.g. [`binarytrees`](https://github.com/ocaml-bench/sandmark/blob/master/benchmarks/multicore-numerical/binarytrees5_multicore.ml) and [`test_decompress`](https://github.com/ocaml-bench/sandmark/blob/master/benchmarks/decompress/test_decompress_multicore.ml)) to measure and optimize global root scanning implementation. It will be useful to know how representative they are of the workloads that saw an improvement with incremental scanning. Happy to make changes to the benchmarks or add new ones to get them closer to real workloads.

**Comment by lpw25** [comment]
Those benchmarks seem to only use C global roots, however it is `caml_globals` that are marked incrementally on trunk and whose performance is likely to be important.

**Comment by Sudha247** [comment]
Yes, that's true, the benchmarks use only C roots and I agree that performance of `caml_globals` is going to be important. C roots have an API with which we could control the number of roots registered and removed. Is there a way to do something similar for native `caml_globals`? If not, are there any reference programs that tend to create a large number of native globals which I could use as a starting point?

**Comment by lpw25** [comment]
I think the obvious candidate would be a program with 100,000 modules in it. That might be a pain to make though, so perhaps you can make a module with 100,000 top-level `ref None`s? I think that they should end-up as each being their own `caml_global` but I could definitely be wrong.

**Comment by chambart** [comment]
The ref None would not do the job here. At least with closure, there is a single root which is the module. But with flambda this should work. Any Initialize_symbol is a root. You could even do that without allocating by doing a lot of `let vn = Sys.opaque_identity 42`

**Comment by gadmm** [comment]
We (@tizoc, @gasche, @jhjourdan and I, working on OCaml/Rust interoperability) have been working the past two weeks on an efficient implementation of an interface similar to `caml_root`. We have a prototype with benchmarks showing good performance, in fact greater than the generational global roots. We have not yet officially announced anything but we thought you would have liked to hear about it. Its description with benchmarks is available here: https://gitlab.com/ocaml-rust/ocaml-boxroot. 

One aspect that is relevant to the current discussion is that we expect that (with some work) it can be made efficiently multithreaded, because we use standard allocation techniques for which the problem is well-studied, in fact it is very similar to the Streamflow allocator from which `runtime/shared_heap.c` is inspired (but organised to scan the roots efficiently). Our main motivation is interoperability with Rust, where our roots will replace both local roots, and global roots when they are used to store OCaml data inside foreign data structures. One problem we need to tackle is cross-domain deallocation of roots, and more generally deallocation of a root from a thread that does not hold any domain lock.

Presumably, the same problems will arise when trying to make global roots efficient, because one common usage of global roots found in the wild (opam) is to malloc a value-sized block, register its contents as a global root, and use it in similar fashion as we intend to with our roots. Assuming per-domain skiplists for global roots to avoid a global lock, either global roots might eventually need similar solutions for cross-domain deallocation, or we could look into making our approach a suitable replacement for this kind of use-case.

So far we have implemented it as a library using OCaml's scanning hooks facility, but for working on latency we have been hitting a wall since here are currently no hooks for scanning incrementally. In the future, we will have to either look into extending those hooks, or experimenting with it directly in the compiler codebase.

**Comment by Sudha247** [comment]
Thanks for the ideas @lpw25 and @chambart, I'll take a look at this in the coming days and keep this thread updated with improvements in benchmark code and numbers.

@gadmm Thank you for bringing this up, I will have a closer look at `ocaml-boxroot` and get back to you if there are any questions/comments.

**Comment by avsm** [comment]
Just adding in some related work as I did some background reading on GC roots: this 2018 paper by @let-def https://arxiv.org/pdf/1812.04905.pdf

**Comment by Sudha247** [comment]
To update the current status of this:
* The bugs listed above have been fixed in #426.
* `caml_root` API was removed in #516, all `caml_root` values are registered as generational global roots.
* The minor collector and the major collector scan global roots exclusively from one domain per cycle. See #463, #472, #567.

I'm closing this one since the bugs are fixed now and `caml_root` API is gone. There's still scope for performance improvements, that are described in #578
